### PR TITLE
Initial MH implementation

### PR DESCRIPTION
**Pull request by cpfiffer**
This PR provides access to a basic implementation of random walk Metropolis-Hastings. Currently it depends on Turing, but only to access the interface API. Once the API has stabilized in Turing we can move it to MCMCChains and clear out the dependencies, but for right now I'm fairly pleased with how lightweight this implementation is.

Usage: 

```julia
# Import the package.
using AdvancedMH

# Define the components of a basic model.
insupport(θ) = θ[2] >= 0
dist(θ) = Normal(θ[1], θ[2])
density(data, θ) = insupport(θ) ? sum(logpdf.(dist(θ), data)) : -Inf

# Generate a set of data from the posterior we want to estimate.
data = rand(Normal(0, 1), 30)

# Construct a DensityModel.
model = DensityModel(density, data)

# Set up our sampler with initial parameters.
spl = MetropolisHastings([0.0, 0.0])

# Sample from the posterior.
chain = sample(model, spl, 100000; param_names=["μ", "σ"])
```

Output:

```julia
Object of type Chains, with data of type 100000×3×1 Array{Float64,3}

Iterations        = 1:100000
Thinning interval = 1
Chains            = 1
Samples per chain = 100000
internals         = lp
parameters        = μ, σ

2-element Array{MCMCChains.ChainDataFrame,1}

Summary Statistics

│ Row │ parameters │ mean     │ std      │ naive_se    │ mcse       │ ess     │ r_hat   │
│     │ Symbol     │ Float64  │ Float64  │ Float64     │ Float64    │ Any     │ Any     │
├─────┼────────────┼──────────┼──────────┼─────────────┼────────────┼─────────┼─────────┤
│ 1   │ μ          │ -0.23908 │ 0.187608 │ 0.00059327  │ 0.00310081 │ 3225.02 │ 1.00003 │
│ 2   │ σ          │ 0.98903  │ 0.138437 │ 0.000437777 │ 0.00242748 │ 2830.8  │ 1.0003  │

Quantiles

│ Row │ parameters │ 2.5%     │ 25.0%     │ 50.0%     │ 75.0%     │ 97.5%    │
│     │ Symbol     │ Float64  │ Float64   │ Float64   │ Float64   │ Float64  │
├─────┼────────────┼──────────┼───────────┼───────────┼───────────┼──────────┤
│ 1   │ μ          │ -0.61185 │ -0.360519 │ -0.236016 │ -0.119666 │ 0.134902 │
│ 2   │ σ          │ 0.758477 │ 0.887894  │ 0.973777  │ 1.07983   │ 1.29455  │
```

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+version = "0.0.1"
+
+[deps]
+Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
+MCMCChains = "c7f686f2-ff18-58e9-bc7b-31028e88f75d"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+Turing = "fce5fe82-541a-59a6-adf8-730c64b5f9a0"
+Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
```
Comment (yebai) [inline Project.toml]: Maybe consider remove the dependency on `Zygote`? Since MH sampler does not require gradient information

File: src/AdvancedMH.jl [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+# Default constructors.
+MetropolisHastings(init_θ::T) where T<:Real = MetropolisHastings{T}(init_θ, x -> Normal(x, 1.0))
+MetropolisHastings(init_θ::T) where T<:Vector{<:Real} = MetropolisHastings{T}(init_θ, x -> MvNormal(x, 1.0))
+
+# Define a model type. Stores the log density function and the data to 
+# evaluate the log density on.
+struct DensityModel{V<:VariateForm, S<:ValueSupport, T} <: Sampleable{V, S}
+    π :: Function
+    data :: T
```
Comment (yebai) [inline src/AdvancedMH.jl]: Maybe 

- consider removing `data`, and make `π` should be a self-contained `logpdf` (i.e. `π` ∝ prior x likelihood)
- change `π` to `ℓπ` (i.e. `logdensity`)

File: src/AdvancedMH.jl [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+# Store the new draw and its log density.
+Transition(model::M, θ::T) where {M<:DensityModel, T} = Transition(θ, ℓπ(model, θ))
+
+# Tell the interface what transition type we would like to use.
+transition_type(model::DensityModel, spl::MetropolisHastings) = typeof(Transition(spl.init_θ, ℓπ(model, spl.init_θ)))
+
+# Define a function that makes a basic proposal depending on a univariate
+# parameterization or a multivariate parameterization.
+proposal(spl::MetropolisHastings, model::DensityModel, θ::Real) = Transition(model, rand(spl.proposal(θ)))
```
Comment (yebai) [inline src/AdvancedMH.jl]: `Real` type is slow. Consider changing it to parametric type, e.g. 

`proposal(spl::MetropolisHastings, model::DensityModel, θ::T) where T<:Real`

File: src/AdvancedMH.jl [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+proposal(spl::MetropolisHastings, model::DensityModel, θ::Vector{<:Real}) = Transition(model, rand(spl.proposal(θ)))
+proposal(spl::MetropolisHastings, model::DensityModel, t::Transition) = proposal(spl, model, t.θ)
+
+# Calculate the logpdf of one proposal given another proposal.
+q(spl::MetropolisHastings, θ1::Real, θ2::Real) = logpdf(spl.proposal(θ1), θ2)
```
Comment (yebai) [inline src/AdvancedMH.jl]: Slightly hard to follow at first glance. Maybe add a comment: `spl.proposal(θ1)` returns a distribution object representing the proposal.

File: src/AdvancedMH.jl [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+) where {
+    M <: DensityModel,
+    S <: MetropolisHastings,
+    T <: Transition
+}
+    # Generate a new proposal.
+    θ = proposal(spl, model, θ_prev)
+    
+    # Calculate the log acceptance probability.
+    α = ℓπ(model, θ) - ℓπ(model, θ_prev) + q(spl, θ_prev, θ) - q(spl, θ, θ_prev)
```
Comment (yebai) [inline src/AdvancedMH.jl]: There might be a mistake/typo here. The MH correct accept ratio is 

`π(θ)q(θ_prev | θ) / π(θ_prev)q(θ | θ_prev)`

which translate to 

`ℓπ(model, θ) - ℓπ(model, θ_prev) - q(spl, θ_prev, θ) + q(spl, θ, θ_prev)`

File: src/AdvancedMH.jl [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    # Decide whether to return the previous θ or the new one.
+    if log(rand()) < min(α, 0.0)
```
Comment (yebai) [inline src/AdvancedMH.jl]: Consider passing in `rng::AbstractRNG` here since it calls random generator.

**Comment by yebai** [comment]
Thanks, @cpfiffer! Nice work - I left some comments above. 

In addition, I think we can also consider removing the dependency on `Distributions`, which is a big dependency just for bringing in `Normal` and `MvNormal`. But this is a less urgent item and can be done in a separate effort. 

@xukai92 @trappmartin @willtebbutt Do you have any comments?

File: src/AdvancedMH.jl [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+import MCMCChains: Chains
+import Turing.Interface: step!, AbstractSampler, AbstractTransition, transition_type
+
+# Exports
+export MetropolisHastings, DensityModel, sample
+
+# Define a sampler type.
+struct MetropolisHastings{T} <: AbstractSampler 
+    init_θ :: T
+    proposal :: Function
```
Comment (trappmartin) [inline src/AdvancedMH.jl]: Why does this have to be a Function? I find this confusing, especially as you always access the proposal distribution the function returns. Can we change this to a Distribution? Or is the idea that this could be a NN, e.g. normalising flow (but that's in case of Bijectors again a Distributions).

File: src/AdvancedMH.jl (hunk 3 above)
Comment (willtebbutt) [reply src/AdvancedMH.jl]: > Real type is slow. Consider changing it to parametric type, e.g.

No, this is fine here. `Real` is only slow if used as a field of a struct, it's fine as a type constraint as an argument to a function.

> Also, maybe rename proposal to propose?

Agree with this.

File: src/AdvancedMH.jl (hunk 7 above)
Comment (willtebbutt) [reply src/AdvancedMH.jl]: I agree that restricting this to be a `Function` is a bit odd, and agree that requiring it to be a distribution of some kind might be more appropriate. Alternatively, we could not restrict at all for now, until we're sure what we actually want. e.g.
```julia
struct MetropolisHastings{Tinit, Tproposal} <: AbstractSampler
    init_θ::Tinit
    proposal::Tproposal
end
```

File: src/AdvancedMH.jl [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+end
+
+# Default constructors.
+MetropolisHastings(init_θ::T) where T<:Real = MetropolisHastings{T}(init_θ, x -> Normal(x, 1.0))
+MetropolisHastings(init_θ::T) where T<:Vector{<:Real} = MetropolisHastings{T}(init_θ, x -> MvNormal(x, 1.0))
```
Comment (willtebbutt) [inline src/AdvancedMH.jl]: Similarly
```suggestion
MetropolisHastings(init_θ::Vector{<:Real}) = MetropolisHastings(init_θ, x -> MvNormal(x, 1.0))
```

File: src/AdvancedMH.jl (hunk 2 above)
Comment (willtebbutt) [reply src/AdvancedMH.jl]: Also, please consider making the type of `ℓπ` parametric for type stability reasons.

File: src/AdvancedMH.jl [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+# Define a model type. Stores the log density function and the data to 
+# evaluate the log density on.
+struct DensityModel{V<:VariateForm, S<:ValueSupport, T} <: Sampleable{V, S}
+    π :: Function
+    data :: T
+end
+
+# Default density constructor.
+DensityModel(π::Function, data::T) where T = DensityModel{VariateForm, ValueSupport, T}(π, data)
```
Comment (willtebbutt) [inline src/AdvancedMH.jl]: Why do we need this?

File: src/AdvancedMH.jl [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+# Create a very basic Transition type, only stores the 
+# parameter draws and the log probability of the draw.
+struct Transition{T<:Union{Vector{<:Real}, <:Real}} <: AbstractTransition
+    θ :: T
+    lp :: Float64
```
Comment (willtebbutt) [inline src/AdvancedMH.jl]: Please consider making parametric

File: src/AdvancedMH.jl [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    model::M,
+    spl::S,
+    N::Integer;
+    kwargs...
+) where {
+    M <: DensityModel,
+    S <: MetropolisHastings
+}
+    return Transition(model, spl.init_θ)
+end
```
Comment (willtebbutt) [inline src/AdvancedMH.jl]: This can probably be re-written to
```julia
function step!(
    rng::AbstractRNG,
    model::DensityModel,
    spl::MetropolisHastings,
    N::Integer,
    kwargs...,
)
    return Transition(model, spl.init_θ)
end
```
since the `where` doesn't add anything.

File: src/AdvancedMH.jl [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+# either a new proposal (if accepted) or the previous proposal 
+# (if not accepted).
+function step!(
+    rng::AbstractRNG,
+    model::M,
+    spl::S,
+    ::Integer,
+    θ_prev::T;
+    kwargs...
+) where {
```
Comment (willtebbutt) [inline src/AdvancedMH.jl]: Same here. No need for the `where`

File: src/AdvancedMH.jl [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    end
+
+    # Add the log density field to the parameter names.
+    push!(param_names, "lp")
+
+    # Bundle everything up and return a Chains struct.
+    return Chains(vals, param_names, (internals=["lp"],))
+end
+
+end
```
Comment (willtebbutt) [inline src/AdvancedMH.jl]: Extra empty line?

File: test/runtests.jl [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    # Set up our sampler.
+    spl = MetropolisHastings([0.0, 0.0], x -> MvNormal(x, 1.0))
+
+    # Sample from the posterior.
+    chain = sample(model, spl, 100000; param_names=["μ", "σ"])
+
+    # chn_mean ≈ dist_mean atol=atol_v
+    @test mean(chain["μ"].value) ≈ 0.0 atol=0.1
+    @test mean(chain["σ"].value) ≈ 1.0 atol=0.1
+end
```
Comment (willtebbutt) [inline test/runtests.jl]: Missing empty line?

**Comment by willtebbutt** [comment]
Thanks for the excellent PR @cpfiffer . I've left a number of obnoxious comments :)

File: src/AdvancedMH.jl (hunk 7 above)
Comment (xukai92) [reply src/AdvancedMH.jl]: I'm a bit confused here. From the example, what should be passed in is `x -> MvNormal(x, 0.5)`. It's like a distribution construcotor, or proposal constructor? Instead of an actual proposal distribution. Is my understanding correct?

File: src/AdvancedMH.jl (hunk 5 above)
Comment (cpfiffer) [reply src/AdvancedMH.jl]: Numerically what I was doing was accurate, but the way I had built `q` was very strange, since `q(spl, θ_prev, θ) = q(θ | θ_prev)`, and not the other way as might be expected. I fixed `q` and corrected the arrangement of the accept ratio.

File: src/AdvancedMH.jl (hunk 7 above)
Comment (cpfiffer) [reply src/AdvancedMH.jl]: I'm not sure what the alternative is here -- how would you dynamically construct the conditional proposal distribution if the proposal wasn't a function? In this case I'm aiming for RWMH, which requires a new proposal distribution for each parameter draw. There is not (to my knowledge) an easy way to pass a general distribution that can be reparameterized when taking conditionals. Functions make this easy and fast. 

The only alternative I see still requires the generation of an anonymous function, and it would be significantly less general than the functional requirement.

File: src/AdvancedMH.jl (hunk 9 above)
Comment (cpfiffer) [reply src/AdvancedMH.jl]: We need this to pass in the abstract types `VariateForm` and `ValueSupport`, because otherwise users would have to annoyingly specify it themselves. Most people are unlikely to care, so this is just a convenience function. This is part of allowing models to be subtypes of `Sampleable`, which we haven't really though too hard about -- it's kind of left in the interface just in case we want to step the modelling language up to the next level.

File: src/AdvancedMH.jl (hunk 7 above)
Comment (xukai92) [reply src/AdvancedMH.jl]: Just to be clear.
- I agree it should be a function.
    - We need to parametrise the struct with function as Will suggests though.
- The name `proposal` might not be accurate? I would feel what being constructed is the proposal, but not the function itself.

File: src/AdvancedMH.jl (hunk 10 above)
Comment (xukai92) [inline src/AdvancedMH.jl]: Are we going to generalise this somehow to store more information?

Comment (cpfiffer) [reply src/AdvancedMH.jl]: The generalization comes from adding fields for the specific information you want to store or adding a new `Transition` type. It'll be expanded as we store more things.

**Comment by trappmartin** [comment]
Thanks, looks really cool. 
And thanks for adding comments that explains what the field `proposal` is.

File: src/AdvancedMH.jl (hunk 5 above)
Comment (yebai) [reply src/AdvancedMH.jl]: Ah, I see. Thanks for the clarification. Maybe consider separating `q` into two steps for better clarity, i.e.:  

- `q(θ, κ(spl, θ_prev)` 

here `κ` constructs a proposal kernel from `θ_prev` then compute the probability of  `θ` given this kernel.

File: src/AdvancedMH.jl (hunk 7 above)
Comment (yebai) [reply src/AdvancedMH.jl]: Maybe consider separating the proposal into 

- `θ = θ_prev + Δθ`

where `Δθ` is sampled from the `proposal` distribution.  The `propose` function needs to be modified so that it returns `θ + Δθ` as output. This way we can fix the `proposal`, with potential efficiency improvements.

**Comment by cpfiffer** [comment]
I've updated this so that the proposal distribution is fixed. No more functional weirdness. It also uses the currently unregistered packaged [AbstractMCMC](https://github.com/TuringLang/AbstractMCMC.jl/), which will soon be the future home of all the MCMCChain interface stuff that currently sits in Turing.

**Comment by cpfiffer** [comment]
I've updated this so that the proposal distribution is fixed. No more functional weirdness. It also uses the currently unregistered packaged [AbstractMCMC](https://github.com/TuringLang/AbstractMCMC.jl/), which will soon be the future home of all the MCMCChain interface stuff that currently sits in Turing.
